### ISSUE TITLE
Original full lang value not available within translation functions

### ISSUE THREAD

**Issue by xdev1** [shoelace-style/localize#6]
Let's assume you have a non-trivial parametrized translation where the translation function itself is using some localization functions for example for date or number formatting, how does the translation function know the full current `lang` value?
It's just a rhetorical question, of course, as the full `lang` value is not retrievable within the translation function as it is not passed here:

https://github.com/shoelace-style/localize/blob/62d5a9ed31d5b4a891b6b466a4be896e20599cbe/src/index.ts#L66

In general, when you provide the translations, you know at least the base lang value (for example 'en') but not necessarily the full lang value (for example 'en-ZA'):

Some possible solutions:

1. Pass the `lang` argument as first argument to the local `term` function:
    `return term(lang, ...args);`
    Obvious disadvantage: That `lang` argument will not be used in > 90% of all cases, yet you have to write `lang: string, ` all the time, which may feel a bit annoying.

2. Don't use varargs here (frankly, I've never been a fan of this varargs solution here in the first place), but pass the parameters as object (with properly named keys) as first argument and the full `lang` value as second. My personally preferred solution.

Passing something as `this` to the translation function would be possible but ugly as hell, I guess, so I have not listed it as possible solution.
[EDIT] And I also would not like passing the parameters as tuple/array.

**Comment by claviska** [comment]
> how does the translation function know the full current lang value?

Each translation [is a module](https://github.com/shoelace-style/shoelace/blob/next/src/translations/en.ts), so if you needed to reference it, you can set a variable and reuse it throughout the translation. In the context of a component, [you can now use `lang()` and `dir()`](https://github.com/shoelace-style/localize/commit/d5dc62824103d393647603a9515a7bc2587052d2) to access those properties as of the latest version.

Does that cover your use case?

**Comment by xdev1** [comment]
I guess my description above was not very understandable, I'll try to explain it with an example:

Be aware that for example the number formats for `en-US` and `en-ZA` differ:
en-US: 12,345.67
en-ZA: 12 345,67

Now, see the the following example from `localize`'s `README.md`

```ts
const translation: Translation = {
  ...
  
  // Plurals
  num_files_selected: (count: number) => {
    if (count === 0) return 'No files selected';
    if (count === 1) return '1 file selected';
    return `${count} files selected`;
  }
};

```

In the example `count` normally will be less than 1000.
But in those cases where numbers can be greater than 1000 or may be decimals it is often requested to also localize that number values.

```ts
// en.ts
import type { number, Translation } from '@shoelace-style/localize';

const translation: Translation = {
  ...
  
  // Plurals
  // this is currently not supported by the localize library
  num_files_selected: (params: { count: number }, lang: string) => {
    if (params.count === 0) return 'No files selected';
    if (params.count === 1) return '1 file selected';
    return `${number(lang, params.count)} files selected`;
  }
};
```

The second version will work out of the box properly for 'en', 'en-US' and 'en-ZA' and all other English speaking locales. With the current version of `@shoelace-style/localize` library you would need an extra `en-ZA.ts` localization module. Things are getting even worse if you use date formatting within a translation function, then you would need extra translation modules for `en-GB`, `en-AU`, `en-ZA` and maybe most other Commonwealth countries, which would be completely unpractical.

I hope, now it's clear what I mean.

**Comment by xdev1** [comment]
Of course you can always pass the full `lang` value explicitly, like in 
```ts
  this.localize.term('someTermKey', someParam, this.localize.lang())
```
But that that would be annoying, wouldn't it?

What about passing the `lang` value implicitly as additional argument?

Instead of
https://github.com/shoelace-style/localize/blob/62d5a9ed31d5b4a891b6b466a4be896e20599cbe/src/index.ts#L66

just using

```ts
   return term(...args, lang) as string; 
```

**Comment by claviska** [comment]
I still don't see why this is a problem. From inside the language file, you can do this:

```ts
import { registerTranslation } from '../utilities/localize';
import type { Translation } from '../utilities/localize';

const code = 'en-GB';

const translation: Translation = {
  $code: code,
  $name: 'English',
  $dir: 'ltr',

  clearEntry: 'Clear entry',
  close: 'Close',
  copy: 'Copy',
  currentValue: 'Current value',
  hidePassword: 'Hide password',
  progress: 'Progress',
  remove: 'Remove',
  resize: 'Resize',
  scrollToEnd: 'Scroll to end',
  scrollToStart: 'Scroll to start',
  selectAColorFromTheScreen: 'Select a color from the screen',
  showPassword: 'Show password',
  toggleColorFormat: 'Toggle color format'
};

registerTranslation(translation);

export default translation;
```

And anywhere inside the translation, you have access to `code`. Outside of the translation, the new `lang` method can be used to access it.

Am I missing a use case?

**Comment by xdev1** [comment]
How do you make the following English translation module work properly for the US and ALL (English speaking) Commonwealth countries (be aware that both date format and language format differ in English speaking countries) WITHOUT providing separate language modules for all 54 Commonwealth countries?
Be aware that both the date value and the number value shall be localized properly.

```js
const translation: Translation = {
  $code: 'en',
  $name: 'English',
  $dir: 'ltr',

  dailyCovidCases: (day: Date, count: number) => {
    // TODO: Localize ${day} and ${number}
    //  If you use 'en' as `lang` then you'll have the US date and
    // number format also in for example en_ZA which is just wrong!
    return "Number of COVID cases on ${day}: ${number}";
  }
};
```

**Comment by xdev1** [comment]
Thanks for the response. Frankly, I'm not really convinced.
You gave the translators a lot of power by allowing them the use of functions for translations (which was a brilliant idea, IMHO). But with great power comes great responsibility. With or without access to the "real" `lang` value, the translators can do an infinite number of stupid things within these translation functions.
And if we made the `LocalizeController` a bit more universal (= removing `implements ReactiveController` and the methods `hostConnected` and `hostDisconnected` and moving that logic to the constructor, plus making the constructor a bit more powerful and supportive for other use cases) and passing a `LocalizeController` instance to the translations functions implicitly (instead of just `lang` as suggested above), we could even get rid of all other exported function except for `registerTranslation` and `LocalizeController`, which would be quite nice and would simplify the public API, IMHO (at least I think we would not need the other functions any longer). The whole stuff would have more or less the same code size as the current version (actually, I've implemented a POC yesterday, that's why I know).

BTW: i18next is basically doing something similar as I've suggested:
https://codesandbox.io/s/demo1-ssqlof?file=/src/index.ts

I will close this issue then. TYVM for your time.
